### PR TITLE
Add batch saga update APIs to speed up scheduler kill jobs

### PR DESCRIPTION
**Pull request by dgdonut**
The scheduler's processing of kill job requests in the case of jobs with large numbers of unstarted tasks is very bad - O(N) where N is the number of unstarted tasks. For each N, the scheduler applies separate start and end task saga updates, each of which independently write-lock the Saga data structure as well as the entire in-memory saga log. This is now O(1). When unstarted task counts are >~ 10,000 this results in a perf increase of over two orders of magnitude, as show by go test -bench of the added Benchmark.

I will resist substantial cleanup or refactor of the saga or other APIs until we are able to dedicate the time to properly overhaul them.

### CODE REVIEW

File: saga/saga.go [hunk 1]
```diff
@@ -196,20 +213,25 @@ type sagaUpdate struct {
...
+		if err != nil {
+			return oldState, err
+		}
 	}
 
-	//try durably storing the message
-	err = log.LogMessage(msg)
+	// try durably storing the message
+	if len(msgs) == 1 {
+		err = log.LogMessage(msgs[0])
```
Comment (dgdonut) [inline saga/saga.go:224]: just using LogBatchMessages breaks like 20 million mock tests and I am refusing to fix them

File: scheduler/server/stateful_scheduler_test.go [hunk 2]
```diff
@@ -594,6 +592,32 @@ func Test_StatefulScheduler_GetSomeThrottledStatus(t *testing.T) {
...
+		duration: 5 * time.Second,
+	}
+	taskDurations["foo"].update(21 * time.Second)
+	taskDurations["foo"].update(25 * time.Second)
+	if taskDurations["foo"].duration != 17*time.Second {
+		t.Fatalf("Expected 17 seconds, got %v", taskDurations["foo"].duration)
+	}
+}
+
+func BenchmarkProcessKillJobsRequests(b *testing.B) {
```
Comment (ryancouto) [inline scheduler/server/stateful_scheduler_test.go:608]: `*testing.B`
oooh cool

File: saga/saga.go [hunk 3]
```diff
@@ -144,24 +149,36 @@ func (s *Saga) StartCompensatingTask(taskId string, data []byte) error {
 // Returns an error if it fails
 //
 func (s *Saga) EndCompensatingTask(taskId string, results []byte) error {
-	return s.updateSagaState(MakeEndCompTaskMessage(s.id, taskId, results))
+	return s.updateSagaState([]SagaMessage{MakeEndCompTaskMessage(s.id, taskId, results)})
+}
+
+//
```
Comment (ryancouto) [inline saga/saga.go]: extra //

**Review by ryancouto — APPROVED** [review]

File: saga/saga.go [hunk 4]
```diff
@@ -144,24 +149,36 @@ func (s *Saga) StartCompensatingTask(taskId string, data []byte) error {
...
 
 	result := <-resultCh
 
 	// after we successfully log an EndSaga message close the channel
 	// no more messages should be logged
-	if msg.MsgType == EndSaga {
-		close(s.updateCh)
+	for _, msg := range msgs {
+		if msg.MsgType == EndSaga {
+			close(s.updateCh)
```
Comment (JeanetteBruno) [inline saga/saga.go:173]: can we break here?

File: saga/saga.go (hunk 1 above)
Comment (JeanetteBruno) [reply saga/saga.go:224]: can we just remove the call to log.LogMessage() - we aren't really doing an durable store anyway.  (same as above) (and remove the broken tests?)

Comment (dgdonut) [reply saga/saga.go:224]: I don't follow, we are doing a durable store here.

File: saga/saga.go (hunk 3 above)
Comment (dgdonut) [reply saga/saga.go]: removed all of these

File: saga/saga.go (hunk 4 above)
Comment (dgdonut) [reply saga/saga.go:173]: yep, done

File: saga/saga.go (hunk 1 above)
Comment (dgdonut) [reply saga/saga.go:224]: I just spent a while trying to consolidate the mock based tests and it's too painful, deferring it to wider refactor

**Pull request merged**